### PR TITLE
Convert stream database to async/await.

### PR DESCRIPTION
**Pull request by clokep**
Converts the stream database to async/await and adds some type information.

### CODE REVIEW

**Comment by clokep** [comment]
I think I failed to run tests on this before pushing. I'll fix those up...

**Comment by clokep** [comment]
I was able to steal some test fixes from #8071 to fix this. 👍

**Comment by clokep** [comment]
I made an additional fix here which was to move a file from `synapse.storage` to `synapse.api`. It was defining a class (really a `namedtuple`), but by being in storage it caused things to start importing storage classes, which was causing an import loop.

File: synapse/storage/databases/main/stream.py [hunk 1]
```diff
@@ -68,8 +70,12 @@
 
 
 def generate_pagination_where_clause(
-    direction, column_names, from_token, to_token, engine
-):
+    direction: str,
+    column_names: Tuple[str, str],
+    from_token: Optional[Tuple[int, int]],
+    to_token: Optional[Tuple[int, int]],
+    engine,
```
Comment (erikjohnston) [inline synapse/storage/databases/main/stream.py:77]: Can we change this to `Union[PostgresEngine, SqliteEngine]`? It feels a bit wrong to leave it untyped

File: synapse/storage/databases/main/stream.py [hunk 2]
```diff
@@ -132,7 +136,9 @@ def generate_pagination_where_clause(
     return " AND ".join(where_clause)
 
 
-def _make_generic_sql_bound(bound, column_names, values, engine):
+def _make_generic_sql_bound(
+    bound: str, column_names: Tuple[str, str], values: Tuple[Optional[int], int], engine
+):
```
Comment (erikjohnston) [inline synapse/storage/databases/main/stream.py:141]:
```suggestion
) -> str:
```

File: synapse/storage/databases/main/stream.py (hunk 1 above)
Comment (clokep) [reply synapse/storage/databases/main/stream.py:77]: 👍 I think I was afraid this would create crazy import loops.

Comment (clokep) [reply synapse/storage/databases/main/stream.py:77]: Actually, it seems there's a `BaseDatabaseEngine` which I should use instead.

**Pull request merged**